### PR TITLE
ROX-25990: Prevent online/offline status flapping in Sensor

### PR DESCRIPTION
**Pull request by vikin91**
### TL;DR

- (the fix) Go online first when SyncFinished event happens - emit `SensorComponentEventCentralReachable` event shortly before SyncFinished.
- To not confuse the future us that gRPC connection is up and ready:
- Adjust comments and log messages about the connection state
- Rename introduce `SensorComponentEventCentralReachableHTTP` that is emitted in the same place where `SensorComponentEventCentralReachable` used to happen (it is unused now, but it should help us understand better how the connection is established.

🔍 💬 Can be reviewed commit-by-commit, although I made one revert of an idea that I didn't like, so maybe reviewing normally is a better option.

### Description

This PR **prevents** Sensor going online too early. In some circumstances, Sensor may assume that the gRPC connection is healthy and go online, whereas the first gRPC call may fail and the state will be reset to offline. That may lead to online/offline states flapping several times before the connection stabilizes.

The exemplary consequence of switching between offline and online too frequently (more frequently than once per minute) may be: connections and process indicators reported from collector not being enriched and thus lost forever.

This PR is not fixing the flake that is reported in ROX-25990. Instead, it uses the knowledge acquired when working on the flake and fixes the faulty behavior outside of the integration tests - in particular in the e2e tests and the production runs. 

Sensor integration tests use fake gRPC connection factory that is different to the connection factory used in reality. This means that the issue being fixed in the production code may not cause the integration tests to behave correctly. For that to happen, we would need to fix the same issue in the code that is used only in the integration test.

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

Existing e2e test should report problems if the connection between Sensor and Central would not work.

#### How I validated my change

- We cannot rely on integration tests for this particular change, as the integration tests use different code branch
- Manually on a cluster
   - [x] Happy path
   - [x] Cycling online-offline-online and ensuring that the transitions of those states in components look ok (by looking at the logs) 
   - [x] Using various toxiproxy configs trying to break the gRPC connection and making sure that the connection recovers and sensor goes online when toxic profiles are removed.

### CODE REVIEW

File: sensor/common/sensor/sensor.go [hunk 1]
```diff
@@ -411,6 +416,7 @@ func (s *Sensor) notifySyncDone(syncDone *concurrency.Signal, centralCommunicati
 	case <-syncDone.Done():
 		s.currentStateMtx.Lock()
 		defer s.currentStateMtx.Unlock()
+		s.changeStateNoLock(common.SensorComponentEventCentralReachable)
 		s.notifyAllComponents(common.SensorComponentEventSyncFinished)
```
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/sensor/sensor.go:420]: Well, that is an idea to consider... However, I am not fully sure yet whether a merge would be the correct thing to do (but also not saying that it is incorrect). Here is my train of thoughts:

1. Ping to Central is a false-friend, as it is done over HTTP. We can use it as a necessary condition for the central communication to be up, but it is not a sufficient condition - I saw few cases when ping was okay, but gRPC had problems.
2. SyncFinished is a valuable information - we know that there is a working connection to Central AND an important operation of sync has finished. Sync is an indication that connection is okay, but a signal that connection is okay is not an indication that sync finished.
3. What I actually meant with this change was to mark Central as Reachable as soon as the first communication over gRPC connection finishes with success. SyncFinished was a comfortable thing to pick, but actually a much better moment would be when the Hello protocol is finished with success.

So, I would rather suggest to keep the Sync Finished as it is and maybe add another event `HelloComplete` and go online immediately after it? We could also name the `HelloComplete` as `CentralReachable`, although we may think of better name to avoid confusion with HTTP-Central-Ping.

The idea is great and we should do this, but I would like to avoid merging states when we may loose some information about the Sensor state. Does that make sense? :)

File: sensor/common/sensor/sensor.go [hunk 2]
```diff
@@ -299,6 +299,7 @@ func (s *Sensor) Start() {
 	if features.PreventSensorRestartOnDisconnect.Enabled() {
 		log.Info("Running Sensor with connection retry: preventing sensor restart on disconnect")
 		go s.communicationWithCentralWithRetries(&centralReachable)
+		// There is no guarantee that the connection to central will be ready when this go routine finishes!
```
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/sensor/sensor.go:302]: This func `communicationWithCentralWithRetries` is kind of a special thing :) 
It does 3 things:
- checks the status to maybe change online/offline mode
- triggers the connection
- waits for connection to be closed (and then retries) or waits for a stop signal (and then does not retry).

So this
>  If the routine finishes, then we know for sure the connection is stopped as it only stops if sensor stops.

is not 100% correct, cause we would need to check the signals additionally to tell whether that is an intended or unintended stop (for the former we do not try to reconnect, but for the latter we do retry).

[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/sensor/sensor.go:302]: > is not 100% correct, cause we would need to check the signals additionally to tell whether that is an intended or unintended stop (for the former we do not try to reconnect, but for the latter we do retry).

🤔 Maybe I misunderstood something but AFAIU the routine can only finish if the `stop` signal is triggered. If that happens, we call `centralCommunication.Stop` before exiting the routine. So if the routine has ended two things have/are happened/happening: Sensor is stopping and the connection with Central is stopped. If my understanding is correct this:

```golang
// There is no guarantee that the connection to central will be ready when this go routine finishes!
```

is not correct because when the go routine finishes, it stops the connection.

I think we might be saying the same thing, but I think the comment is kind of confusing because the way I'd read it without knowing how `communicationWithCentralWithRetries` works is: this go routine is meant to finish when the connection is established or failed. When in reality it does not stop (unless sensor stops).

File: sensor/common/sensor/sensor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/sensor/sensor.go:420]: I think it all boils down to what the components need to know, or in other words: 

- Does any component need to know when Central is reachable (as in ping succeeded)? 
- Does any component need to know when the hello message is successfully sent?

If the answer is no to both, then adding a new state for `HelloComplete` is not really worth it IMO and I'd remove the `CentralReachable` status. Just keep `SyncFinished` and switch the components to online then.

If the answer is yes to the first question, then I'd keep the `CentralReachable` where it is right now (right after the ping) and I'd change the components that do not need to react to `CentralReachable` to react to `SyncFinished` instead.

If the answer is yes to the second question, then I'd add (like you suggested) a new state `HelloComplete` and make the necessary components react to it.

Alternatively if we see Sensor as a state machine with components that plug into it, we could provide the three states regardless of whether the components use them.

[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/sensor/sensor.go:420]: Your analysis makes sense, but to provide answers to those two questions, we would need to go over the list of components and check the details for each of them. We may do this, but I am not sure this should be a prerequisite for this fix.

The problems is that `CentralReachable` checks if Central is Reachable over HTTP, whereas the components rely on Central being reachable over gRPC. There is a possibility that Central is reachable over HTTP but not over gRPC. So my proposal boils down to introduction of `CentralReachableGRPC` that would be emitted shortly after the Hello protocol completes. The Hello protocol is not strictly required here - Central maybe gRPC reachable earlier but we would need to write some more code to find that moment - but it is convenient, as  we know that Hello runs over gRPC and when Hello completes, then gRPC must be up and running.

[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/sensor/sensor.go:420]: So for now, the situation looks like this:

1. Central responds to HTTP ping - `CentralReachable`
2. Central finishes Hello protocol - no signal
3. Sync starts - no signal
4. Sync finishes - `SyncFinished`

This is how we could make it correct:

1. Central responds to HTTP ping - `CentralReachableHTTP`
2. Central finishes Hello protocol - `CentralReachableGRPC`
3. Sync starts - no signal
4. Sync finishes - `SyncFinished`

but for convenience, we may use `SyncFinished` as `CentralReachableGRPC` and save us some code.
If we find a scenario where some components could benefit from starting sooner than `SyncFinished`, we may add the `CentralReachableGRPC`.

For now, I will rely on `SyncFinished` for simplicity. 
However, I would like rename the `CentralReachable` to `CentralReachableHTTP` to avoid confusion.

File: sensor/common/sensor/sensor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vikin91) [reply sensor/common/sensor/sensor.go:302]: Yeah, now I see that I was wrong. Of course the retry loop can finish in the 3 situations that I listed above, but the goroutine finishes only when we stop retrying and that happens only when we intentionally call `Stop`.

I will adapt the comment to make it correct.

File: sensor/common/centralclient/grpc_connection.go [hunk 3]
```diff
@@ -81,10 +83,11 @@ func (f *centralConnectionFactoryImpl) SetCentralConnectionWithRetries(conn *uti
...
-	// waits until central is ready and has a valid license, otherwise it kills sensor by sending a signal
+	// Waits until central is ready and has a valid license, otherwise it kills sensor by sending a signal.
+	// This ping runs over HTTP and does check the health of gRPC connection.
 	if err := f.pingCentral(); err != nil {
-		log.Errorf("checking central status failed: %v", err)
-		f.stopSignal.SignalWithError(errors.Wrap(err, "checking central status failed"))
+		log.Errorf("checking central status over HTTP failed: %v", err)
+		f.stopSignal.SignalWithError(errors.Wrap(err, "checking central status over HTTP failed"))
 		return
 	}
```
Comment (vikin91) [inline sensor/common/centralclient/grpc_connection.go:93]: Actually, `SensorComponentEventCentralReachableHTTP` could be emitted already in this line.

File: sensor/common/centralclient/grpc_connection.go [hunk 4]
```diff
@@ -118,5 +123,5 @@ func (f *centralConnectionFactoryImpl) SetCentralConnectionWithRetries(conn *uti
 
 	conn.Set(centralConnection)
 	f.okSignal.Signal()
```
Comment (vikin91) [inline sensor/common/centralclient/grpc_connection.go:125]: This emits `SensorComponentEventCentralReachableHTTP` which is not fully correct, as the state is: "central reachable over HTTP" AND "gRPC connection set up but not ready yet".

However, I would like to keep it as it is for now, as in the code proposed in this PR, we stop relying on the `SensorComponentEventCentralReachableHTTP` signal.

**Comment by vikin91** [comment]
Converting to draft as apparently some tests are failing in the CI and I need to give it a look.

File: sensor/common/sensor/sensor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/sensor/sensor.go]: > 1. Central responds to HTTP ping - CentralReachableHTTP
> 2. Central finishes Hello protocol - CentralReachableGRPC
> 3. Sync starts - no signal
> 4. Sync finishes - SyncFinished

I agree with this, the only thing I'm not sure is the usability ATM of `CentralReachableHTTP/GRPC`. I don't think that currently we have components that would need those signals, but it's true that we would have to look at all of them to really know for sure and, it is also true that, we might have components that might need those signals in the future.

**Review by lvalerom — APPROVED** [review]
LGTM. I think we should try to differentiate between `CentralReachable` and `SyncFinished` because right now they're virtually the same signal. But I think that can be address in a follow-up. It would be also interesting to go through the different components and try to understand which signals they need. I think that right now we have a mix of `CentralReachable` and `SyncFinished`. I know for sure some need `SyncFinished`, but I wonder if some of the components that use `CentralReachable` should be using `SyncFinished` and might act-up if at some point we make the differentiation I mentioned.

**Comment by vikin91** [comment]
Thank you for the review!

> we should try to differentiate between CentralReachable and SyncFinished because right now they're virtually the same signal

Fully agree here, but I tried it to some degree ([here](https://github.com/stackrox/stackrox/pull/12989)) and it looks it is not trivial.

>  It would be also interesting to go through the different components and try to understand which signals they need. 

Yes, that would be great too! I think, we may even find one or two that could benefit from Central being reachable over HTTP.

>  I wonder if some of the components that use CentralReachable should be using SyncFinished and might act-up if at some point we make the differentiation I mentioned.

I think, that may have been the case in this PR and that lead me to do the revert and falling back to the separation of the signals. I described it more in the private Slack message, where I explain why - most probably - the integration tests were failing when we had both signals `EventCentralReachable` and `EventResourceSyncFinished` merged and handled as single `EventResourceSyncFinished`.

(The Slack message is obviously not included here, but I am not sure it is relevant for this PR. If some people from the future would require the details from this Slack message, then it means that we messed up something with the events we emit and we should revisit those carefully.)
